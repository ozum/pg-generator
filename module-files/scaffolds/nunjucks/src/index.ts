console.warn("This is a pg-generator plugin. It is not intended for direct usage. Use it vie pg-generator."); // eslint-disable-line no-console
