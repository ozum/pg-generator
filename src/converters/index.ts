export * from "./mermaid-to-svg";
