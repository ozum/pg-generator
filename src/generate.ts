/* eslint-disable no-nested-ternary */
import type { Options } from "./types";
import { getDb, composeWith } from "./utils";

/**
 * Executes the default sub-generator `app` from a generator.
 *
 * @param generator is the name or path of the generator. If it is a local path, use `require.resolve` or provide an absolute path.
 * @param options are the options for the generator.
 *
 * @example
 * generate("generator-from-npm", options);
 * generate(require.resolve("./local-generator"), options);
 */
export async function generate(generator: string, options?: Options): Promise<void>;
/**
 * Executes a sub-generator from a generator.
 *
 * @param generator is the name or path of the generator. If it is a local path, use `require.resolve` or provide an absolute path.
 * @param subGenerator is the name of the generator.
 * @param options are the new options added on top of curent options.
 *
 * @example
 * generate("generator-from-npm", "sub-generator");
 * generate(require.resolve("./local-generator"), "sub-generator", options);
 */
export async function generate(generator: string, subGenerator?: string, options?: Options): Promise<void>;
export async function generate(generator: string, subOrOptions?: string | Options, maybeOptions?: Options): Promise<void> {
  const options = typeof maybeOptions === "object" ? maybeOptions : typeof subOrOptions === "object" ? subOrOptions : {};
  const generatorName = typeof subOrOptions === "string" ? subOrOptions : undefined;
  const db = await getDb(options);
  return composeWith(generator, generatorName, options, db, process.cwd());
}
