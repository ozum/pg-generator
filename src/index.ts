export { PgGenerator } from "./pg-generator";
export { generate } from "./generate";
export { scaffold } from "./scaffold";

// Generator Helpers
export * as filterFunctions from "./utils/filter-functions";
export * as converters from "./converters";

// Types
export type { Context, GeneratorOptions, ClientOptions, Options } from "./types/index";
