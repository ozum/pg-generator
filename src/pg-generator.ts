/* eslint-disable class-methods-use-this */
import { extname, join, dirname, resolve } from "path";
import { promises as fs } from "fs";
import { chalk } from "meow-helper";
import merge from "lodash.merge";
import JSON5 from "json5";
import prettier from "prettier";
import { PgenError } from "./utils/pgen-error";
import type { Context, GeneratorOptions, InternalOptions } from "./types/index";
import { getDbObjects, resolvePath, augmentContext, composeWith, parseTemplatePath, scanTemplateDir } from "./utils";

const COLORS: Record<string, keyof typeof chalk> = {
  default: "green",
  delete: "yellow",
};

/**
 * Base abstract class for for pg-generator classes.
 */
export abstract class PgGenerator<O extends GeneratorOptions = GeneratorOptions> {
  #destinationRoot: string;
  #internalOptions: InternalOptions;
  protected fs: any;
  protected options: O & { envPrefix: string; context: Record<string, any> };

  /**
   * Creates an instance of PgGenerator.
   *
   * @param options are parameters for several options.
   * @param internalOptions are added by `pg-generator` functions and not for user.
   */
  public constructor(options: O, internalOptions: InternalOptions) {
    if (options.clear && !options.outDir) throw new PgenError("As a precaution, 'clear' option cannot be used without 'outDir' option.");
    const collisions = internalOptions.db.relationNameCollisions;
    if (collisions) throw PgenError.collisionError(collisions);

    this.options = { envPrefix: "DB", copy: true, context: {}, ...options };
    this.#internalOptions = internalOptions;
    this.#destinationRoot = resolve(internalOptions.cwd, this.options.outDir || "");
  }

  /**
   * Renders all templates in `[rootDir]/templates` directory with the render function using related context data,
   * and writes generated contents to the files in output directory. If render function returns `undefined`
   * for a template/database object pair no file is not written for that pair.
   *
   * Template file names may contain variables and basic filter functions to change names of generated files.
   *
   * Additionally copies all files in `[rootDir]/files` to the output directory.
   */
  public async generate(): Promise<void> {
    if (this.options.clear) await this.clear();
    await this.init();

    const { templates, files } = await scanTemplateDir(this.#internalOptions.templateDir);
    const context = await this.fetchContext();

    // Generate templates. Provide absolute outdir. Don't copy files, because you should use Yeoman's `mem-fs-editor` API.
    const processTemplates = Promise.all(
      templates.flatMap((path) => {
        const { className, accessor, targetPath } = parseTemplatePath(path);
        return getDbObjects(this.#internalOptions.db, className, accessor, path).map(async (dbObject) => {
          const destinationPath = resolvePath(targetPath, dbObject); // e.g. {name#dashcase}.js -> member-option.js
          const content = await this.format(destinationPath, await this.render(path, augmentContext(dbObject, context)));
          if (content !== undefined && content.replace(/\s/g, "") !== "") await this.process(destinationPath, content);
        });
      })
    );
    const copyFiles = Promise.all(files.map((file) => this.copyTemplate(file, file)));
    await Promise.all([processTemplates, copyFiles]);
  }

  //
  // ────────────────────────────────────────────────────────── I ──────────
  //   :::::: L I F E C Y C L E : :  :   :    :     :        :          :
  // ────────────────────────────────────────────────────────────────────
  //
  // Methods below are lifecycle methods and most probably would be extended
  // by child classes for custom behavior. Providing `render` is mandatory.
  // Flow: clear -> init --> context -> render -> format -> process

  /**
   * If clear option is true, this method is called. It deletes destination path.
   * User may override this method to tailor clear operation according to generator's
   * purposes.
   */
  protected async clear(): Promise<any> {
    await this.deleteDestination();
  }

  /**
   * Executed after clear operation.
   */
  protected init(): Promise<any> | any {
    return undefined;
  }

  /**
   * Provides generator specific extra context to templates. This data is deeply merged with
   * context provided by user.
   *
   * @returns generator spesific extra context data.
   */
  protected context(): Promise<Record<string, any>> | Record<string, any> {
    return {};
  }

  /**
   * Generates content from the template with provided path using the context.
   *
   * @param templatePath is the path of the template file to generate content.
   * @param context is the data to be used for generating content.
   * @returns generated content.
   */
  protected render(templatePath: string, context: Context): Promise<undefined | string> | undefined | string {
    const stub = templatePath ?? context; // eslint-disable-line @typescript-eslint/no-unused-vars
    return undefined;
  }

  /**
   * Formats given content. By default `prettier` is used. Plase note that at this stage file is not written to disk yet.
   *
   * @param destination is the path of the destination file to be created.
   * @param content is the content to format.
   * @returns formatted code.
   */
  protected format(destination: string, content?: string): Promise<undefined | string> | undefined | string {
    if (content === undefined || content === null || content === "") return content;
    return prettier.format(content, { filepath: destination });
  }

  /**
   * Writes generated content from [[render]] to the destination file.
   *
   * @param destination is the destionation path of the file to be cerated..
   * @param content is the content to process.
   */
  protected process(destination: string, content: string): Promise<any> | any {
    return this.writeTemplate(destination, content);
  }

  //
  // ────────────────────────────────────────────────────── I ──────────
  //   :::::: U T I L I T Y : :  :   :    :     :        :          :
  // ────────────────────────────────────────────────────────────────
  //

  // Protected methods below are utility methods to be used by child classes.

  /**
   * Logs given messsage if `options.log` is true.
   *
   * @param title is the title of the log.
   * @param message is the nessage to log.
   * @param options are options.
   */
  protected log(title: string, { message, source, destination }: { message?: string; source?: string; destination?: string }): void {
    if (!this.options.log) return;
    const titleColor = COLORS[title] ?? COLORS.default;
    const colorizeTitle = chalk[titleColor] as (input: string) => string;
    const logTitle = colorizeTitle(title);
    const messageSpace = message && source && destination ? ": " : " ";
    const logMessage = message === undefined ? "" : ` ${message}${messageSpace}`;
    const logSource = source ? `${source}${destination ? " ➔ " : ""}` : "";
    const logDestination = destination ? join(this.options.outDir || "", destination) : "";
    console.log(`   ${logTitle} ${logMessage}${logSource}${logDestination}`); // eslint-disable-line no-console
  }

  /**
   * Returns abosulte path for the given path relative to output directory.
   *
   * @param path is the path relative to output directory.
   * @returns the absolute path.
   */
  protected destinationPath(path = ""): string {
    return resolve(this.#destinationRoot, path);
  }

  /**
   * Returns abosulte path for the given path relative to template root.
   *
   * @param path is the path relative to template root.
   * @returns the absolute path.
   */
  protected templatePath(path = ""): string {
    return resolve(this.#internalOptions.templateDir, path);
  }

  //
  // ──────────────────────────────────────────────────────── I ──────────
  //   :::::: I N T E R N A L : :  :   :    :     :        :          :
  // ──────────────────────────────────────────────────────────────────
  //

  // Private methods below are for internal usage.

  /**
   * Executes the default sub-generator `app` from a generator. If options are provided, they are combined with current options.
   *
   * @param generator is the name or path of the generator. If it is a local path, use `require.resolve` or provide an absolute path.
   * @param options are the new options added on top of curent options.
   *
   * @throws error if generator can not be found.
   *
   * @example
   * this.composeWith("generator-from-npm");
   * this.composeWith("generator-from-npm", options);
   * this.composeWith(require.resolve("./local-generator"), options);
   */
  protected async composeWith(generator: string, options?: GeneratorOptions): Promise<void>;
  /**
   * Executes a sub-generator from a generator. If options are provided, they are merged with current options. Also some of the options
   * are changed to sensible default values for sub-generators (e.g. `clear` is changed to `false`), but they can be overridden by
   * newly provided options.
   *
   * @param generator is the name or path of the generator. If it is a local path, use `require.resolve` or provide an absolute path.
   * @param subGeneratorName is the name of the sub-generator to be executed.
   * @param options are the new options added on top of curent options.
   *
   * @throws error if generator or sub-generator can not be found.
   *
   * @example
   * this.composeWith("generator-from-npm", "sub-generator");
   * this.composeWith(require.resolve("./local-generator"), "sub-generator", options);
   */
  protected async composeWith(generator: string, subGeneratorName?: string, options?: GeneratorOptions): Promise<void>;
  protected async composeWith(generator: string, subOrOptions?: string | GeneratorOptions, maybeOptions?: GeneratorOptions): Promise<void> {
    const [options = {}, generatorName] = typeof subOrOptions === "string" ? [maybeOptions, subOrOptions] : [maybeOptions];
    const newOptions = { ...this.options, clear: false, ...options };
    return composeWith(generator, generatorName, newOptions, this.#internalOptions.db, this.#internalOptions.cwd);
  }

  /**
   * Writes given content to the file after creating it's directory.
   *
   * @param destinantion is the path to write file.
   * @param content is the content to write.
   */
  private async writeTemplate(destination: string, content: string): Promise<void> {
    const destinationPath = this.destinationPath(destination);
    await fs.mkdir(dirname(destinationPath), { recursive: true });
    await fs.writeFile(destinationPath, content, { encoding: "utf8" });
    this.log("create", { destination });
  }

  /**
   * Copy the source file to the destination. Also creates the destination directory if it does not exist.
   *
   * @param source is the source file to copy.
   * @param destination is the destination file to be created.
   */
  private async copyTemplate(source: string, destination: string): Promise<void> {
    await fs.mkdir(dirname(this.destinationPath(destination)), { recursive: true });
    await fs.copyFile(this.templatePath(source), this.destinationPath(destination));
    this.log("create", { destination });
  }

  /**
   * Removes file or directory (and it's contents() from destination. If destination is equal to `cwd()`
   * it throws error and aborts operation for a safety measure. Most probably `cwd()` contains other files too.
   *
   * @param path is the path relative to the destination root.
   */
  private async deleteDestination(path = ""): Promise<void> {
    const destination = this.destinationPath(path);
    const isRoot = resolve(this.#internalOptions.cwd) === resolve(destination);
    if (isRoot) throw new PgenError("Root is not deleted, because it may be a mistake. Probably it would contain other files.");

    try {
      await fs.rmdir(this.destinationPath(path), { recursive: true });
    } catch (error) {
      if (error.code !== "ENOENT") throw error;
      else if (error.code === "ENOTDIR") await fs.unlink(this.destinationPath(path));
    }
    this.log("delete", { destination: path || "." });
  }

  /**
   * Fetches and merges context data from `pg-generator` instance, context file and options.
   * The context file may be a JSON5 data file (may contain comments etc.) or a JavaScript file
   * with a default export of object or a function. If it is a function, it is passed
   * `pg-structure` db instance.
   *
   * Merge priorities are: `options.context` > context file > `context()` method.
   *
   * @param contextFile is the file to read from.
   * @param context is the context data.
   * @returns merged context data.
   */
  private async fetchContext(): Promise<Record<string, any>> {
    let contextFromFile = {};

    if (this.options.contextFile !== undefined) {
      const contextFile = resolve(this.options.contextFile);
      const fileContent =
        extname(contextFile) === ".json"
          ? JSON5.parse(await fs.readFile(contextFile, { encoding: "utf8" }))
          : (await import(contextFile)).default;
      contextFromFile = typeof fileContent === "function" ? fileContent(this.#internalOptions.db) : fileContent;

      this.log("context", { destination: contextFile });
    }

    return merge(await this.context(), contextFromFile, this.options.context);
  }
}
