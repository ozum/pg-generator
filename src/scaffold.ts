import { join, basename, extname } from "path";
import { EOL } from "os";
import { promises as fs } from "fs";
import { ignoreCode } from "ignor";
import { PgenError } from "./utils/pgen-error";
import { copyDir } from "./utils";

const SCAFFOLD_DIR = join(__dirname, "../module-files/scaffolds");

/**
 * Renames "name" entry of the "package.json" in given directory.
 *
 * @param dir is the directory that contains the "package.json" file.
 * @param newName is the new name.
 */
async function renameInPackageJson(dir: string, newName: string): Promise<void> {
  const pkgPath = join(dir, "package.json");
  const pkg = JSON.parse(await fs.readFile(pkgPath, { encoding: "utf-8" }));
  pkg.name = newName;
  await fs.writeFile(pkgPath, JSON.stringify(pkg, undefined, 2));
}

/**
 * Formats given list of scaffold names.
 *
 * @param scaffolds are the list of the scaffold names.
 * @internal
 */
export function getScaffoldsMessage(scaffolds: string[]): string {
  const available = scaffolds.join(`${EOL}  ● `);
  return `Available scaffolds are:${EOL}  ● ${available}`;
}

/**
 * Read list of scaffolds from disk.
 *
 * @internal
 */
export async function readScaffolds(): Promise<string[]> {
  return (await fs.readdir(SCAFFOLD_DIR, { withFileTypes: true })).filter((e) => e.isDirectory()).map((e) => e.name);
}

/**
 * Creates a skeleton Node.js project for a generator (pg-generator plugin).
 *
 * @param scaffoldName is the name to create scaffold from.
 * @param options are options
 * @param options.outDir is the output directory to create the project.
 * @param options.name is the name of the created project.
 * @throws error if required scaffold is unavailable or output directory contains files.
 */
export async function scaffold(
  scaffoldName: string,
  { outDir, name = `pg-generator-${basename(outDir, extname(outDir)).replace("pg-generator-", "")}` }: { outDir: string; name?: string }
): Promise<void> {
  const scaffolds = await readScaffolds();
  if (scaffoldName === undefined) throw new PgenError(`SCAFFOLD is required. ${getScaffoldsMessage(scaffolds)}`);
  if (!scaffolds.includes(scaffoldName)) throw new PgenError(`Unknown scaffold "${scaffold}". ${getScaffoldsMessage(scaffolds)}`);

  try {
    const destinationEntries = await fs.readdir(outDir).catch(ignoreCode("ENOENT"));
    if (destinationEntries && destinationEntries.length > 0) throw new PgenError(`Output directory "${outDir}" is not empty.`);
  } catch (error) {
    if (error.code === "ENOTDIR") throw new PgenError(`"${outDir}" is not a directory.`);
    throw error;
  }

  await copyDir(join(SCAFFOLD_DIR, scaffoldName), outDir);
  await renameInPackageJson(outDir, name);
}
