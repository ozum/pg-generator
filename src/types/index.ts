import type { DbObject, Db, Options as PgStructureOptions } from "pg-structure";
import type { Client, ClientConfig } from "pg"; // eslint-disable-line import/no-unresolved

/** Options for generation and reverse engineering process. Options extends [pg-structure options](https://www.pg-structure.com/nav.02.api/interfaces/options) */
export interface GeneratorOptions {
  /** Whether to clear the destination directory before creating files. */
  clear?: boolean;
  /** Path of the output directory. */
  outDir?: string;
  /** Path to a JSON or JS file providing extra context for templates. */
  contextFile?: string;
  /** Extra context data. This data is merged with and overridden by data from context file. */
  context?: Record<string, any>;
  /** Whether to log output to console. */
  log?: boolean;
}

export interface ClientOptions extends PgStructureOptions {
  /** Either a [node-postgres client](https://node-postgres.com/api/client) or a configuration object or a connection string to create a [node-postgres client](https://node-postgres.com/api/client). */
  client?: Client | ClientConfig | string;

  // ─── PG CLIENT OPTIONS ──────────────────────────────────────────────────────────

  /** Database username. Default from environment var: PGUSER || USER || DB_USER */
  user?: string;
  /** Database password. Default from environment var: PGPASSWORD || DB_PASSWORD */
  password?: string;
  /** Database host. Default from environment var: PGHOST || DB_HOST */
  host?: string;
  /** Database to connect. Default from environment var: PGDATABASE || DB_DATABASE */
  database?: string;
  /** Database port. Default from environment var: PGPORT || DB_PORT */
  port?: string;
  /** Connection string to connect to the database e.g. postgres://user:password@host:5432/database */
  connectionString?: string;
  /** Passed directly to node.TLSSocket, supports all tls.connect options */
  ssl?: any;
}

/** Extended options which includes internal options added by `pg-generator` not by user. */
export interface InternalOptions {
  /** The path of the directory that contains template files. */
  templateDir: string;
  /** The path of the cwd . If outpur directory is a relative path, it is calculated relative to destination root. */
  cwd: string;
  /** [pg-structure Db instance](https://www.pg-structure.com/nav.02.api/classes/db)  */
  db: Db;
}

export type Options = GeneratorOptions & ClientOptions;

/** Context provided to render function. */
export type Context = { o: Db | DbObject; x: Record<string, any> };
