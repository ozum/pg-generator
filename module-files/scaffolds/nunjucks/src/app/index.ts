/* eslint-disable class-methods-use-this */
import { PgGenerator, Context } from "pg-generator";
import { getNunjucks } from "../utils/get-nunjucks";
import generatorContext from "./context";

const nunjucks = getNunjucks(__dirname);

/**
 * Example `pg-generator` generator which creates sequelize models.
 */
export default class App extends PgGenerator {
  /**
   * Returns a context data including sequelize data types and some other examples.
   */
  protected context(): Record<string, any> {
    return generatorContext;
  }

  /**
   * Render method which uses a nunjucks template. You can use any template engine.
   * `pg-generator` is template engine agnostic.
   */
  protected async render(templatePath: string, context: Context): Promise<string> {
    return nunjucks.render(templatePath, context);
  }
}
