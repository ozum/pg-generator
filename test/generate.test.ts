import { join } from "path";
import { promises as fs, existsSync } from "fs";
import { load } from "fs-structure";
import { ignoreCode } from "ignor";
import { tmpdir } from "os";
import { generate, Options } from "../src/index";
import App from "./test-helper/pg-generator-test/src/app/index";

const TEST_ROOT = join(tmpdir(), "pgen");
const outDir = join(TEST_ROOT, "out");
const outDir2 = join(TEST_ROOT, "out2");
const app = new App({ outDir }, { db: {}, templateDir: join(__dirname, "templates") } as any);
const CLIENT_OPTIONS = { database: "pg-generator-test", user: "user", password: "password" };

// WORKAROUND: Jest does not output error of dynamic import in (src/utils/compose-with.ts). This function rethrows error.
async function g(subGenerator = "app", options?: Options, { argLength = 3, cwd = "" } = {}): ReturnType<typeof load> {
  const oldCwd = process.cwd();
  if (cwd) process.chdir(cwd);

  const mergedOptions = {
    clear: true,
    log: false,
    outDir,
    relationNameFunctions: "optimal",
    ...CLIENT_OPTIONS,
    ...options,
  };

  try {
    // To test generate function with 2 paramerets and 3 parameters.
    if (argLength === 3) await generate(join(__dirname, "test-helper/pg-generator-test/src"), subGenerator, mergedOptions);
    else if (argLength === 2) await generate(join(__dirname, "test-helper/pg-generator-test/src", subGenerator), mergedOptions);
    else if (argLength === 1) await generate(join(__dirname, "test-helper/pg-generator-test/src", subGenerator));
    const structure = await load(mergedOptions.outDir ?? outDir);
    process.chdir(oldCwd);
    return structure;
  } catch (error) {
    process.chdir(oldCwd);
    throw new Error(error.message);
  }
}

afterAll(async () => {
  await fs.rmdir(TEST_ROOT, { recursive: true });
});

describe("generate", () => {
  it("should generate files.", async () => {
    expect(await g("app")).toMatchSnapshot();
  });

  it("should generate files if called with 2 arguments.", async () => {
    expect(await g("app", {}, { argLength: 2 })).toMatchSnapshot();
  });

  it("should generate files if called with 1 argument.", async () => {
    await expect(() => g("app", {}, { argLength: 1, cwd: outDir })).rejects.toThrow(" have same names");
  });

  it("should add context from context module using options.", async () => {
    expect(await g("app", { contextFile: join(__dirname, "test-helper/context.ts") })).toMatchSnapshot();
  });

  it("should add context from context module exporting function using options.", async () => {
    expect(await g("app", { contextFile: join(__dirname, "test-helper/context-as-function.ts") })).toMatchSnapshot();
  });

  it("should add context from context JSON file using options.", async () => {
    expect(await g("app", { contextFile: join(__dirname, "test-helper/context.json") })).toMatchSnapshot();
  });

  it("should not create files if render returns undefined.", async () => {
    expect(await g("useless")).toMatchSnapshot();
  });

  it("should throw error for relation name collisions.", async () => {
    await expect(() => g("app", { relationNameFunctions: "short" })).rejects.toThrow("relations have same names");
  });

  it("should throw error if clear is true without outDir.", async () => {
    await expect(() => g("app", { outDir: undefined })).rejects.toThrow("without 'outDir' option");
  });

  it("should not clear output directory if clear is false.", async () => {
    await g("app", { clear: false, outDir: outDir2 });
    expect(existsSync(outDir2)).toBe(true);
  });

  it("should throw if target is cwd and clear is true", async () => {
    await expect(() => g("app", { outDir: process.cwd() })).rejects.toThrow("CWD cannot be deleted");
  });

  it("should compose.", async () => {
    // TODO: eslint yada prettier snapshot'ı formatlıyor olabilir mi. Normalde çalışan test release'de hata veriyor.
    // lint-staged'den lint ve formatı kaldırınca hata olmuyor. test/__snapshots__ dizinini exclude et.
    expect(await g("composer")).toMatchSnapshot();
  });

  it("should output cwd if no outDir is provided.", async () => {
    await fs.rmdir(outDir, { recursive: true }).catch(ignoreCode("ENOENT"));
    await fs.mkdir(outDir);
    expect(await g("app", { outDir: undefined, clear: false }, { cwd: outDir })).toMatchSnapshot();
  });

  it("should throw if a filter in file name cannot be found.", async () => {
    await expect(() => g("error-path-filter")).rejects.toThrow("There is no 'non-existing' filter");
  });

  it("should throw if file name contains non-existing db object type.", async () => {
    await expect(() => g("error-db-object-type")).rejects.toThrow("'NonExisting' is not a known database object");
  });

  it("should throw if file name contains non-existing db object attribute.", async () => {
    await expect(() => g("error-db-object-attribute")).rejects.toThrow("'none' cannot be found");
  });

  it("should throw if generator cannot be found.", async () => {
    await expect(() => generate("XYZ")).rejects.toThrow("You don't seem to have a generator with the name 'XYZ' installed.");
  });

  it("should throw if sub-generator cannot be found.", async () => {
    await expect(() => g("XYZ")).rejects.toThrow("doesn't seem to have a sub-generator with the name 'XYZ'");
  });
});

describe("App", () => {
  it("should return default destination path.", () => {
    expect(app.defaultDestinationPath()).toBe(outDir);
  });

  it("should return default template path.", () => {
    expect(app.defaultTemplatePath()).toBe(join(__dirname, "templates"));
  });
});
