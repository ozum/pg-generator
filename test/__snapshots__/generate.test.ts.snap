// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate should add context from context JSON file using options. 1`] = `
Object {
  "README.md": "# Hello!
",
  "archive.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "archive/archive.md": "# Schema:
",
  "index.js": "const PurchaseHistory = \\"\\";
const Account = \\"\\";
const Contact = \\"\\";
const LineItem = \\"\\";
const Product = \\"\\";
const Purchase = \\"\\";
const PurchaseHistory = \\"\\";
const Region = \\"\\";
const SystemOption = \\"\\";
const SystemOptionHistory = \\"\\";
",
  "public.account.js": "id;
name;
primary_contact_id;
secondary_contact_id;
",
  "public.contact.js": "id;
account_id;
region_Id;
name;
surname;
phone;
preferred_languages;
priority;
",
  "public.line-item.js": "purchase_id;
product_id;
quantity;
unit_price;
",
  "public.product.js": "id;
name;
price;
",
  "public.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "public.purchase.js": "id;
created_at;
purchased_at;
contact_id;
",
  "public.region.js": "id;
name;
parentid;
",
  "public.system-option-history.js": "id;
system_option_key;
system_option_sub_key;
",
  "public.system-option.js": "key;
sub_key;
value;
",
  "public/public.md": "# Schema:
",
}
`;

exports[`generate should add context from context module exporting function using options. 1`] = `
Object {
  "README.md": "# Hello!
",
  "archive.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "archive/archive.md": "# Schema:
",
  "index.js": "const PurchaseHistory = \\"\\";
const Account = \\"\\";
const Contact = \\"\\";
const LineItem = \\"\\";
const Product = \\"\\";
const Purchase = \\"\\";
const PurchaseHistory = \\"\\";
const Region = \\"\\";
const SystemOption = \\"\\";
const SystemOptionHistory = \\"\\";
",
  "public.account.js": "id;
name;
primary_contact_id;
secondary_contact_id;
",
  "public.contact.js": "id;
account_id;
region_Id;
name;
surname;
phone;
preferred_languages;
priority;
",
  "public.line-item.js": "purchase_id;
product_id;
quantity;
unit_price;
",
  "public.product.js": "id;
name;
price;
",
  "public.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "public.purchase.js": "id;
created_at;
purchased_at;
contact_id;
",
  "public.region.js": "id;
name;
parentid;
",
  "public.system-option-history.js": "id;
system_option_key;
system_option_sub_key;
",
  "public.system-option.js": "key;
sub_key;
value;
",
  "public/public.md": "# Schema:
",
}
`;

exports[`generate should add context from context module using options. 1`] = `
Object {
  "README.md": "# Hello!
",
  "archive.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "archive/archive.md": "# Schema:
",
  "index.js": "const PurchaseHistory = \\"\\";
const Account = \\"\\";
const Contact = \\"\\";
const LineItem = \\"\\";
const Product = \\"\\";
const Purchase = \\"\\";
const PurchaseHistory = \\"\\";
const Region = \\"\\";
const SystemOption = \\"\\";
const SystemOptionHistory = \\"\\";
",
  "public.account.js": "id;
name;
primary_contact_id;
secondary_contact_id;
",
  "public.contact.js": "id;
account_id;
region_Id;
name;
surname;
phone;
preferred_languages;
priority;
",
  "public.line-item.js": "purchase_id;
product_id;
quantity;
unit_price;
",
  "public.product.js": "id;
name;
price;
",
  "public.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "public.purchase.js": "id;
created_at;
purchased_at;
contact_id;
",
  "public.region.js": "id;
name;
parentid;
",
  "public.system-option-history.js": "id;
system_option_key;
system_option_sub_key;
",
  "public.system-option.js": "key;
sub_key;
value;
",
  "public/public.md": "# Schema:
",
}
`;

exports[`generate should compose. 1`] = `
Object {
  "README.md": "# Hello!
",
  "archive.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "archive/archive.md": "# Schema:
",
  "index.js": "const ArchivePurchaseHistory = \\"\\";
const PublicAccount = \\"\\";
const PublicContact = \\"\\";
const PublicLineItem = \\"\\";
const PublicProduct = \\"\\";
const PublicPurchase = \\"\\";
const PublicPurchaseHistory = \\"\\";
const PublicRegion = \\"\\";
const PublicSystemOption = \\"\\";
const PublicSystemOptionHistory = \\"\\";
",
  "public.account.js": "id;
name;
primary_contact_id;
secondary_contact_id;
",
  "public.contact.js": "id;
account_id;
region_Id;
name;
surname;
phone;
preferred_languages;
priority;
",
  "public.line-item.js": "purchase_id;
product_id;
quantity;
unit_price;
",
  "public.product.js": "id;
name;
price;
",
  "public.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "public.purchase.js": "id;
created_at;
purchased_at;
contact_id;
",
  "public.region.js": "id;
name;
parentid;
",
  "public.system-option-history.js": "id;
system_option_key;
system_option_sub_key;
",
  "public.system-option.js": "key;
sub_key;
value;
",
  "public/public.md": "# Schema:
",
  "useless.md": "# Useless
",
}
`;

exports[`generate should generate files if called with 2 arguments. 1`] = `
Object {
  "README.md": "# Hello!
",
  "archive.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "archive/archive.md": "# Schema:
",
  "index.js": "const ArchivePurchaseHistory = \\"\\";
const PublicAccount = \\"\\";
const PublicContact = \\"\\";
const PublicLineItem = \\"\\";
const PublicProduct = \\"\\";
const PublicPurchase = \\"\\";
const PublicPurchaseHistory = \\"\\";
const PublicRegion = \\"\\";
const PublicSystemOption = \\"\\";
const PublicSystemOptionHistory = \\"\\";
",
  "public.account.js": "id;
name;
primary_contact_id;
secondary_contact_id;
",
  "public.contact.js": "id;
account_id;
region_Id;
name;
surname;
phone;
preferred_languages;
priority;
",
  "public.line-item.js": "purchase_id;
product_id;
quantity;
unit_price;
",
  "public.product.js": "id;
name;
price;
",
  "public.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "public.purchase.js": "id;
created_at;
purchased_at;
contact_id;
",
  "public.region.js": "id;
name;
parentid;
",
  "public.system-option-history.js": "id;
system_option_key;
system_option_sub_key;
",
  "public.system-option.js": "key;
sub_key;
value;
",
  "public/public.md": "# Schema:
",
}
`;

exports[`generate should generate files. 1`] = `
Object {
  "README.md": "# Hello!
",
  "archive.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "archive/archive.md": "# Schema:
",
  "index.js": "const ArchivePurchaseHistory = \\"\\";
const PublicAccount = \\"\\";
const PublicContact = \\"\\";
const PublicLineItem = \\"\\";
const PublicProduct = \\"\\";
const PublicPurchase = \\"\\";
const PublicPurchaseHistory = \\"\\";
const PublicRegion = \\"\\";
const PublicSystemOption = \\"\\";
const PublicSystemOptionHistory = \\"\\";
",
  "public.account.js": "id;
name;
primary_contact_id;
secondary_contact_id;
",
  "public.contact.js": "id;
account_id;
region_Id;
name;
surname;
phone;
preferred_languages;
priority;
",
  "public.line-item.js": "purchase_id;
product_id;
quantity;
unit_price;
",
  "public.product.js": "id;
name;
price;
",
  "public.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "public.purchase.js": "id;
created_at;
purchased_at;
contact_id;
",
  "public.region.js": "id;
name;
parentid;
",
  "public.system-option-history.js": "id;
system_option_key;
system_option_sub_key;
",
  "public.system-option.js": "key;
sub_key;
value;
",
  "public/public.md": "# Schema:
",
}
`;

exports[`generate should not create files if render returns undefined. 1`] = `
Object {
  "useless.md": "# Useless
",
}
`;

exports[`generate should output cwd if no outDir is provided. 1`] = `
Object {
  "README.md": "# Hello!
",
  "archive.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "archive/archive.md": "# Schema:
",
  "index.js": "const ArchivePurchaseHistory = \\"\\";
const PublicAccount = \\"\\";
const PublicContact = \\"\\";
const PublicLineItem = \\"\\";
const PublicProduct = \\"\\";
const PublicPurchase = \\"\\";
const PublicPurchaseHistory = \\"\\";
const PublicRegion = \\"\\";
const PublicSystemOption = \\"\\";
const PublicSystemOptionHistory = \\"\\";
",
  "public.account.js": "id;
name;
primary_contact_id;
secondary_contact_id;
",
  "public.contact.js": "id;
account_id;
region_Id;
name;
surname;
phone;
preferred_languages;
priority;
",
  "public.line-item.js": "purchase_id;
product_id;
quantity;
unit_price;
",
  "public.product.js": "id;
name;
price;
",
  "public.purchase-history.js": "id;
contact_id;
purchased_at;
total_amount;
",
  "public.purchase.js": "id;
created_at;
purchased_at;
contact_id;
",
  "public.region.js": "id;
name;
parentid;
",
  "public.system-option-history.js": "id;
system_option_key;
system_option_sub_key;
",
  "public.system-option.js": "key;
sub_key;
value;
",
  "public/public.md": "# Schema:
",
}
`;
